File: server/channels/utils/file.go
// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

package utils

import (
	"errors"
	"io"
)

var ErrSizeLimitExceeded = errors.New("size limit exceeded")

type LimitedReaderWithError struct {
	limitedReader *io.LimitedReader
}

func NewLimitedReaderWithError(reader io.Reader, maxBytes int64) *LimitedReaderWithError {
	return &LimitedReaderWithError{
		limitedReader: &io.LimitedReader{R: reader, N: maxBytes + 1},
	}
}

func (l *LimitedReaderWithError) Read(p []byte) (int, error) {
	n, err := l.limitedReader.Read(p)
	if l.limitedReader.N <= 0 && err == io.EOF {
		return n, ErrSizeLimitExceeded
	}
	return n, err
}

File: e2e-tests/playwright/support/file.ts
// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import path from 'node:path';
import fs from 'node:fs';

const assetPath = path.resolve(__dirname, 'asset');

/**
 * Reads file data and creates a File object.
 * @param filePath - The path to the file.
 * @param mimeType - The MIME type of the file.
 * @returns A File object containing the file data.
 * @throws If the file does not exist.
 */
export function getFileData(filePath: string, mimeType: string): File {
    if (!fs.existsSync(filePath)) {
        throw new Error(`File not found at path: ${filePath}`);
    }

    const fileName = path.basename(filePath);
    const fileBuffer = fs.readFileSync(filePath);

    return new File([fileBuffer], fileName, {type: mimeType});
}

/**
 * Reads file data from the "asset" directory and creates a File object.
 * @param filename - The name of the file in the "asset" directory.
 * @param mimeType - The MIME type of the file.
 * @returns An object containing a File object and the filename.
 */
export function getFileDataFromAsset(filename: string, mimeType: string) {
    const filePath = path.join(assetPath, filename);

    return {file: getFileData(filePath, mimeType), filename: path.basename(filePath)};
}

/**
 * Reads file data and creates a Blob object.
 * @param filePath - The path to the file.
 * @param mimeType - The MIME type of the file.
 * @returns A Blob object containing the file data.
 * @throws If the file does not exist.
 */
export function getBlobData(filePath: string, mimeType: string): Blob {
    if (!fs.existsSync(filePath)) {
        throw new Error(`File not found at path: ${filePath}`);
    }
    const fileBuffer = fs.readFileSync(filePath);

    return new Blob([fileBuffer], {type: mimeType});
}

/**
 * Reads file data from the "asset" directory and creates a Blob object.
 * @param filename - The name of the file in the "asset" directory.
 * @param mimeType - The MIME type of the file.
 * @returns An object containing a Blob object and the filename.
 */
export function getBlobDataFromAsset(filename: string, mimeType: string) {
    const filePath = path.join(assetPath, filename);

    return {blob: getBlobData(filePath, mimeType), filename: path.basename(filePath)};
}
